massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,731 @@
+use thiserror::Error;
+
+use crate::block::Block;
+use crate::block_header::{BlockHeader, BlockHeaderSerializer, SecuredHeader};
+use crate::endorsement::{Endorsement, EndorsementSerializer, SecureShareEndorsement};
+use crate::operation::Operation;
+use crate::slot::{Slot, SlotSerializer};
+
+use massa_hash::Hash;
+use massa_serialization::{OptionSerializer, SerializeError, Serializer, U32VarIntSerializer};
+use massa_signature::{MassaSignatureError, PublicKey, Signature};
+
+//
+
+/// Can we build a Denunciation from this object?
+pub trait Denounceable {
+    /// Return true if object can be denounced, false otherwise
+    fn is_denounceable() -> bool {
+        false
+    }
+
+    /// Return DenunciationData (used to modify Signature for `SecureShare<T>`)
+    /// This is a default impl, return None if Self::is_denounceable() returns false
+    fn get_denunciation_data(&self) -> Option<DenunciationData> {
+        match Self::is_denounceable() {
+            true => unimplemented!(),
+            false => None,
+        }
+    }
+}
+
+/// Denunciation data (to be include in `SecureShare<T>` signature
+pub enum DenunciationData {
+    ///
+    Endorsement((Slot, u32)),
+    ///
+    BlockHeader(Slot),
+}
+
+/// A Serializer for `DenunciationData`
+#[derive(Default)]
+pub struct DenunciationDataSerializer {
+    slot_serializer: SlotSerializer,
+    index_serializer: U32VarIntSerializer,
+}
+
+impl DenunciationDataSerializer {
+    /// Create a new `DenunciationDataSerializer`
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Serializer<DenunciationData> for DenunciationDataSerializer {
+    fn serialize(
+        &self,
+        value: &DenunciationData,
+        buffer: &mut Vec<u8>,
+    ) -> Result<(), SerializeError> {
+        match value {
+            DenunciationData::Endorsement(data) => {
+                self.slot_serializer.serialize(&data.0, buffer)?;
+                self.index_serializer.serialize(&data.1, buffer)?;
+            }
+            DenunciationData::BlockHeader(data) => {
+                self.slot_serializer.serialize(data, buffer)?;
+            }
+        };
+        Ok(())
+    }
+}
+
+// Denounceable trait impl
+
+impl Denounceable for Endorsement {
+    fn is_denounceable() -> bool {
```

Comment:
isn't it enough to implem the Denounceable trait to make the object denouncable?

Reply:
Not really because we need to impl this for Operation and Block (because they can be included in SecureShare struct). For these 2 objects, we use the default implentation of the trait